# JSON examples and exercise
****
+ get familiar with packages for dealing with JSON
+ study examples with JSON strings and files 
+ work on exercise to be completed and submitted 
****
+ reference: http://pandas.pydata.org/pandas-docs/stable/io.html#io-json-reader
+ data source: http://jsonstudio.com/resources/
****

In [1]:
import numpy as np
import pandas as pd

## imports for Python, Pandas

In [2]:
import json
from pandas.io.json import json_normalize

## JSON example, with string

+ demonstrates creation of normalized dataframes (tables) from nested json string
+ source: http://pandas.pydata.org/pandas-docs/stable/io.html#normalization

In [3]:
# define json string
data = [{'state': 'Florida', 
         'shortname': 'FL',
         'info': {'governor': 'Rick Scott'},
         'counties': [{'name': 'Dade', 'population': 12345},
                      {'name': 'Broward', 'population': 40000},
                      {'name': 'Palm Beach', 'population': 60000}]},
        {'state': 'Ohio',
         'shortname': 'OH',
         'info': {'governor': 'John Kasich'},
         'counties': [{'name': 'Summit', 'population': 1234},
                      {'name': 'Cuyahoga', 'population': 1337}]}]

****
## JSON example, with file

+ demonstrates reading in a json file as a string and as a table
+ uses small sample file containing data about projects funded by the World Bank 
+ data source: http://jsonstudio.com/resources/

In [4]:
# load json as string
json.load((open('data/world_bank_projects_less.json')))

[{'_id': {'$oid': '52b213b38594d8a2be17c780'},
  'approvalfy': 1999,
  'board_approval_month': 'November',
  'boardapprovaldate': '2013-11-12T00:00:00Z',
  'borrower': 'FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA',
  'closingdate': '2018-07-07T00:00:00Z',
  'country_namecode': 'Federal Democratic Republic of Ethiopia!$!ET',
  'countrycode': 'ET',
  'countryname': 'Federal Democratic Republic of Ethiopia',
  'countryshortname': 'Ethiopia',
  'docty': 'Project Information Document,Indigenous Peoples Plan,Project Information Document',
  'envassesmentcategorycode': 'C',
  'grantamt': 0,
  'ibrdcommamt': 0,
  'id': 'P129828',
  'idacommamt': 130000000,
  'impagency': 'MINISTRY OF EDUCATION',
  'lendinginstr': 'Investment Project Financing',
  'lendinginstrtype': 'IN',
  'lendprojectcost': 550000000,
  'majorsector_percent': [{'Name': 'Education', 'Percent': 46},
   {'Name': 'Education', 'Percent': 26},
   {'Name': 'Public Administration, Law, and Justice', 'Percent': 16},
   {'Name': 'Educatio

****
## JSON exercise

Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,
1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using column 'mjtheme_namecode')
3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

### JSON exercise

Using data in file 'data/world_bank_projects.json' and the techniques demonstrated above,

1. Find the 10 countries with most projects

In [5]:
# load json file into 'json_all'
json_all = pd.read_json('data/world_bank_projects.json')

## Data Exploration 
col = json_all.columns
num_rows = len(json_all)

# print(col)         # explore the column names
# print(num_rows)    # 500 rows of data

# Make new dataframe with columns of our interest and groupby 'countryname' column 
# aggfunc = .count()

new_df = json_all[['countrycode','country_namecode','countryname','countryshortname']]
grouped = new_df.groupby('countryname').count()

# check to see all the columns have the same information across rows by comparing the values in each column across rows
check_equality = (grouped['countrycode'] == grouped['country_namecode']).all()
check_equality_2 = (grouped['countryshortname'] == grouped['country_namecode']).all()

# print statement for readability 
print("Are all the values between 'countrycode' column and 'country_namecode' column same? {}\n\
Are all the values between 'countryshortname' column and 'country_namecode' column same? {}" \
.format(check_equality, check_equality_2))

# Answer (top 10 countries with most projects)
sorted_df = grouped.sort_values(by=['countrycode'], ascending=False)
sorted_df.head(10)



Are all the values between 'countrycode' column and 'country_namecode' column same? True
Are all the values between 'countryshortname' column and 'country_namecode' column same? True


countrycode  country_namecode  \
countryname                                                      
People's Republic of China                19                19   
Republic of Indonesia                     19                19   
Socialist Republic of Vietnam             17                17   
Republic of India                         16                16   
Republic of Yemen                         13                13   
People's Republic of Bangladesh           12                12   
Nepal                                     12                12   
Kingdom of Morocco                        12                12   
Republic of Mozambique                    11                11   
Africa                                    11                11   

                                 countryshortname  
countryname                                        
People's Republic of China                     19  
Republic of Indonesia                          19  
Socialist Republic of Vietnam                  17  
Republic of India                              16  
Republic of Yemen                              13  
People's Republic of Bangladesh                12  
Nepal                                          12  
Kingdom of Morocco                             12  
Republic of Mozambique                         11  
Africa                                         11

### Answer

Top 10 countries with most projects are: 
1. People's Republic of China
2. Republic of Indonesia
3. Socialist Republic of Vietnam
4. Republic of India
5. Republic of Yemen
6. People's Republic of Bangladesh
7. Nepal
8. Kingdom of Morocco
9. Republic of Mozambique
10. Africa

### JSON exercise: 

2. Find the top 10 major project themes (using column 'mjtheme_namecode')

In [6]:
# # Make dataframe out of the json file for debugging purposes
json_df = pd.read_json('data/world_bank_projects.json')

# # Subset dataframe of our interest 
test_df = json_df[['_id','countryname','theme_namecode','mjtheme','mjtheme_namecode','mjthemecode']]
test_df.head()

# # Data Exploration
# cols = json_df.columns
# for i in cols:
#     print(i)

# Column with values in list
# json_df[['mjtheme']]

# # # Main Question
# # flatten out json files using json raw data
with open('data/world_bank_projects.json') as file: 
    data = json.load(file)

# my code takes keyword arguments / example above takes positional arguments 
flat_json = json_normalize(data, record_path='mjtheme_namecode', meta=['countryname'])

# # Data Exploration for flat_json
# len(flat_json['code'].unique())  # 11 unique values (for mjtheme code)
# flat_json[flat_json['code']=='1'].shape  # number of rows for which the 'code' column is '1'

# # Return the top 10 codes 
value_counts = flat_json['code'].value_counts()
value_counts


11    250
10    216
8     210
2     199
6     168
4     146
7     130
5      77
9      50
1      38
3      15
Name: code, dtype: int64

Top 10 major project themes by codes are 

1. 11
2. 10
3. 8
4. 2
5. 6
6. 4
7. 7
8. 5 
9. 9
10. 1


*Major project themes by strings to follow on the next exercise*


### JSON Exercis  

(Partically Combined with Question 2)
3. In 2. above you will notice that some entries have only the code and the name is missing. Create a dataframe with the missing names filled in.

In [46]:

import numpy as np

# # Main Question
# # flatten out json files using json raw data
with open('data/world_bank_projects.json') as file: 
    data = json.load(file)

flat_json = json_normalize(data, record_path='mjtheme_namecode', meta=['countryname'])

# To fill in the missing values of the 'name' column, make a dataframe out of a dictionary and 
# use VLOOKUP-esque method from pandas 

sorted_df = flat_json.sort_values(by=['code','name'],ascending=[True, True])

# 1. Making a dictionary using 'code' column as key and 'name' column as value: 
# method: by zipping two lists together (from the dataframe columns) to avoid using for loops! 
mapping_dict = dict(zip(sorted_df.code, sorted_df.name))

# 2. Create dataframe from this dictionary 
dict_df = pd.DataFrame.from_dict(mapping_dict, orient='index', columns=['name'])

# 3. Use left merge to fill in nan values
merged = sorted_df.merge(dict_df, left_on='code', right_index=True, how='inner')

merged = merged[['code','name_y','countryname']]




# # Return the top 10 codes with code + theme + counts 
value_counts = flat_json['code'].value_counts()
df_value_counts = pd.DataFrame(value_counts).reset_index()
df_value_counts.columns = ['theme_code','counts']
df_value_counts['theme'] = df_value_counts['theme_code'].map(mapping_dict)
# rearranging the columns 
df_final = df_value_counts[['theme_code','theme','counts']]
df_final





theme_code                                         theme  counts
0          11  Environment and natural resources management     250
1          10                             Rural development     216
2           8                             Human development     210
3           2                      Public sector governance     199
4           6         Social protection and risk management     168
5           4      Financial and private sector development     146
6           7                   Social dev/gender/inclusion     130
7           5                         Trade and integration      77
8           9                             Urban development      50
9           1                           Economic management      38
10          3                                   Rule of law      15

### Answer

Top 10 themes are 
1. Environment and natural resources management - 250 times
2. Rural development - 216 times
3. Human development - 210 times
4. Public sector governance	- 199 times
5. Social protection and risk management - 168 times
6. Financial and private sector development - 146 times
7. Social dev/gender/inclusion	- 130 times
8. Trade and integration - 77 times
9. Urban development - 50 times
10. Economic management	- 38 times
